sklearn/cluster/hierarchical.py

```diff
@@ -696,7 +717,15 @@ def fit(self, X, y=None):
                              " instance, got 'memory={!r}' instead.".format(
                                  type(memory)))
 
-        if self.n_clusters <= 0:
+        if (self.distance_threshold and self.n_clusters) or \
+                (not self.n_clusters and not self.distance_threshold):
+            raise ValueError("Either n_clusters (>0) or distance_threshold "
+                             "needs to be set, got n_clusters={} and "
+                             "distance_threshold={} instead. "
+                             "Please set n_clusters=None to continue.".format(
```

Comment:
instad of asking the user to set `n_clusters=None` :
* Set `n_clusters=None` and `distance_threshold=None` by default
* in the `fit` stage, do a validation to check if both are `None`, in that case revert to the previous default behavior of `n_clusters=2`
* Update the docstrings to reflect such behaviour

cc: @jnothman is this ok?